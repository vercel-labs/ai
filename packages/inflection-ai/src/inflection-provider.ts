import {
  LanguageModelV1,
  ProviderV1,
  UnsupportedFunctionalityError,
} from '@ai-sdk/provider';
import {
  FetchFunction,
  loadApiKey,
  withoutTrailingSlash,
} from '@ai-sdk/provider-utils';
import { InflectionChatLanguageModel } from './inflection-chat-language-model';
import {
  InflectionChatModelId,
  InflectionChatSettings,
} from './inflection-chat-settings';

export interface InflectionProvider extends ProviderV1 {
  (
    modelId: InflectionChatModelId,
    settings?: InflectionChatSettings,
  ): LanguageModelV1;

  /**
   * Creates a model for text generation.
   */
  languageModel(
    modelId: InflectionChatModelId,
    settings?: InflectionChatSettings,
  ): LanguageModelV1;

  /**
   * Creates a model for text generation.
   */
  chat(
    modelId: InflectionChatModelId,
    settings?: InflectionChatSettings,
  ): LanguageModelV1;

  /**
   * Not supported by Inflection AI
   * @throws {UnsupportedFunctionalityError}
   */
  textEmbeddingModel(modelId: string): never;
}

export interface InflectionProviderSettings {
  /**
   * Use a different URL prefix for API calls, e.g. to use proxy servers.
   * The default prefix is `https://layercake.pubwestus3.inf7ks8.com/external/api/inference`
   */
  baseURL?: string;

  /**
   * API key that is being sent using the `Authorization` header.
   * It defaults to the `INFLECTION_API_KEY` environment variable.
   */
  apiKey?: string;

  /**
   * Custom headers to include in the requests.
   */
  headers?: Record<string, string>;

  /**
   * Custom fetch implementation. You can use it as a middleware to intercept requests,
   * or to provide a custom fetch implementation for e.g. testing.
   */
  fetch?: FetchFunction;
}

/**
 * Create an Inflection AI provider instance.
 */
export function createInflection(
  options: InflectionProviderSettings = {},
): InflectionProvider {
  const baseURL =
    withoutTrailingSlash(options.baseURL) ??
    'https://layercake.pubwestus3.inf7ks8.com/external/api/inference';

  const getHeaders = () => ({
    Authorization: `Bearer ${loadApiKey({
      apiKey: options.apiKey,
      environmentVariableName: 'INFLECTION_API_KEY',
      description: 'Inflection',
    })}`,
    ...options.headers,
  });

  const createChatModel = (
    modelId: InflectionChatModelId,
    settings: InflectionChatSettings = {},
  ) =>
    new InflectionChatLanguageModel(modelId, settings, {
      provider: 'inflection.chat',
      baseURL,
      headers: getHeaders,
      fetch: options.fetch,
    });

  const provider = function (
    modelId: InflectionChatModelId,
    settings?: InflectionChatSettings,
  ) {
    if (new.target) {
      throw new Error(
        'The Inflection model function cannot be called with the new keyword.',
      );
    }

    return createChatModel(modelId, settings);
  };

  provider.languageModel = createChatModel;
  provider.chat = createChatModel;
  provider.textEmbeddingModel = () => {
    throw new UnsupportedFunctionalityError({
      functionality: 'Text embeddings are not supported by Inflection AI',
    });
  };

  return provider;
}

/**
 * Default Inflection provider instance.
 */
export const inflection = createInflection();
