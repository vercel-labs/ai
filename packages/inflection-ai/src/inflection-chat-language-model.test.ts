import { expect, describe, it, beforeEach } from 'vitest';
import { LanguageModelV1Prompt } from '@ai-sdk/provider';
import {
  createTestServer,
  convertReadableStreamToArray,
} from '@ai-sdk/provider-utils/test';
import { createInflection } from './inflection-provider';

const TEST_PROMPT: LanguageModelV1Prompt = [
  { role: 'user', content: [{ type: 'text', text: 'Hello' }] },
];

const INFERENCE_URL =
  'https://layercake.pubwestus3.inf7ks8.com/external/api/inference';
const STREAMING_URL = `${INFERENCE_URL}/streaming`;

/**
 * Test server for mocking API responses
 */
const server = createTestServer({
  [INFERENCE_URL]: {
    response: {
      type: 'json-value',
      body: {
        created: 1714688002.0557644,
        text: 'Hello there!',
      },
    },
  },
  [STREAMING_URL]: {
    response: {
      type: 'stream-chunks',
      chunks: [
        '{"created": 1728094708.2514212, "idx": 0, "text": "Hello"}',
        '{"created": 1728094708.5789802, "idx": 1, "text": " there"}',
        '{"created": 1728094708.7364252, "idx": 2, "text": "!"}',
      ],
    },
  },
});

beforeEach(() => {
  // Reset the test server before each test
  server.urls[INFERENCE_URL].response = {
    type: 'json-value',
    body: {
      created: 1714688002.0557644,
      text: 'Hello there!',
    },
  };
  server.urls[STREAMING_URL].response = {
    type: 'stream-chunks',
    chunks: [
      '{"created": 1728094708.2514212, "idx": 0, "text": "Hello"}',
      '{"created": 1728094708.5789802, "idx": 1, "text": " there"}',
      '{"created": 1728094708.7364252, "idx": 2, "text": "!"}',
    ],
  };
});

const provider = createInflection({
  apiKey: 'test-api-key',
  baseURL: INFERENCE_URL,
});

const model = provider.chat('inflection_3_pi');

describe('doGenerate', () => {
  it('should extract text response', async () => {
    server.urls[INFERENCE_URL].response = {
      type: 'json-value',
      body: {
        created: 1714688002.0557644,
        text: 'Hello there!',
      },
    };

    const result = await model.doGenerate({
      inputFormat: 'prompt',
      mode: { type: 'regular' },
      prompt: TEST_PROMPT,
    });

    expect(result.text).toBe('Hello there!');
  });

  it('should throw error when trying to use tools', async () => {
    await expect(
      model.doGenerate({
        inputFormat: 'prompt',
        mode: {
          type: 'regular',
          tools: [
            {
              type: 'function',
              name: 'weatherTool',
              description: 'Get weather information',
              parameters: { type: 'object', properties: {} },
            },
          ],
        },
        prompt: TEST_PROMPT,
      }),
    ).rejects.toThrow('Tool calls are not supported by Inflection AI');
  });

  it('should extract usage', async () => {
    server.urls[INFERENCE_URL].response = {
      type: 'json-value',
      body: {
        created: 1714688002.0557644,
        text: 'Hello there!',
      },
    };

    const { usage } = await model.doGenerate({
      inputFormat: 'prompt',
      mode: { type: 'regular' },
      prompt: TEST_PROMPT,
    });

    // Since Inflection API doesn't return usage info, we should expect undefined or estimated values
    expect(usage).toStrictEqual({
      promptTokens: expect.any(Number),
      completionTokens: expect.any(Number),
    });
  });
});

describe('doStream', () => {
  it('should stream text deltas', async () => {
    server.urls[STREAMING_URL].response = {
      type: 'stream-chunks',
      chunks: ['Hello', ' there', '!'],
    };

    const result = await model.doStream({
      inputFormat: 'prompt',
      mode: { type: 'regular' },
      prompt: TEST_PROMPT,
    });

    const parts = await convertReadableStreamToArray(result.stream);
    expect(parts).toMatchSnapshot();
  });
});
