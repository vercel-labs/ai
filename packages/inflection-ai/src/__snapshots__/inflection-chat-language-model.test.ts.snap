// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`doStream > should stream text deltas 1`] = `
[
  {
    "finishReason": "stop",
    "type": "finish",
    "usage": {
      "completionTokens": 0,
      "promptTokens": 2,
    },
  },
]
`;
